### PR TITLE
Fix dynamic label template validation

### PR DESCRIPTION
**Pull request by Konstantinov-Innokentii**
1. Fix https://github.com/grafana/irm/issues/530 - applied same validation logic as for multi-label extraction template to dynamic label.  Actual fix is here - https://github.com/grafana/oncall/pull/5363/files#diff-58657df0f1ff9a8578a14504f1c6cfd240e45e084171c5bbeb09d975c3ec72ddR74
2. Some minor refactorings over static/dynamic/integration label naming. This work should be continued in separate PR.

### CODE REVIEW

File: engine/apps/alerts/models/alert_receive_channel.py [hunk 1]
```diff
@@ -301,16 +301,22 @@ class AlertReceiveChannel(IntegrationOptionsMixin, MaintainableObject):
     rate_limited_in_slack_at = models.DateTimeField(null=True, default=None)
     rate_limit_message_task_id = models.CharField(max_length=100, null=True, default=None)
 
-    AlertGroupCustomLabelsDB = list[tuple[str, str | None, str | None]] | None
-    alert_group_labels_custom: AlertGroupCustomLabelsDB = models.JSONField(null=True, default=None)
     """
-    Stores "custom labels" for alert group labels. Custom labels can be either "plain" or "templated".
-    For plain labels, the format is: [<LABEL_KEY_ID>, <LABEL_VALUE_ID>, None]
-    For templated labels, the format is: [<LABEL_KEY_ID>, None, <JINJA2_TEMPLATE>]
+    alert_group_labels_custom stores config of dynamic labels - mapping of incoming alert payload to labels
```
Comment (Konstantinov-Innokentii) [inline engine/apps/alerts/models/alert_receive_channel.py:305]: Just some renamigs & comments. Further refactoring needed.

File: engine/apps/labels/alert_group_labels.py [hunk 2]
```diff
@@ -99,51 +95,52 @@ def _custom_labels(
         ).only("id", "name")
     }
 
-    rendered_labels = {}
+    result_labels = {}
     for label in alert_receive_channel.alert_group_labels_custom:
-        key_id, value_id, template = label
```
Comment (Konstantinov-Innokentii) [inline engine/apps/labels/alert_group_labels.py:104]: Moved this to separate func

File: engine/apps/labels/alert_group_labels.py [hunk 3]
```diff
@@ -99,51 +95,52 @@ def _custom_labels(
...
-
-    labels = {}
-    for key in rendered_labels:
-        value = rendered_labels[key]
+        label = _apply_dynamic_label_entry(label, label_key_names, label_value_names, raw_request_data)
+        if label:
+            key, value = label
+            result_labels[key] = value
 
-        # check value length
```
Comment (Konstantinov-Innokentii) [inline engine/apps/labels/alert_group_labels.py:129]: Moved to validate_templated_value

File: engine/apps/labels/alert_group_labels.py [hunk 4]
```diff
@@ -99,51 +95,52 @@ def _custom_labels(
...
-        if len(value) > MAX_VALUE_NAME_LENGTH:
-            logger.warning("Template result value is too long. %s", value)
-            continue
+    return result_labels
 
-        labels[key] = value
 
-    return labels
+def _apply_dynamic_label_entry(
+    label: "AlertReceiveChannel.DynamicLabelsEntryDB", keys, values, payload
```
Comment (Konstantinov-Innokentii) [inline engine/apps/labels/alert_group_labels.py:109]: it's old logic of applying custom labels, but without "static" values - they now live in integration associated labels exclusively. Also It calls _validate_templated_value to fix actual big with None values.

**Review by joeyorlando — APPROVED** [review]

**Review by vadimkerr — CHANGES_REQUESTED** [review]

File: engine/apps/alerts/models/alert_receive_channel.py [hunk 5]
```diff
@@ -301,16 +301,23 @@ class AlertReceiveChannel(IntegrationOptionsMixin, MaintainableObject):
...
+    DynamicLabelsEntryDB is a a one entry of a one labels schema. It's a tuple with 3 elements:
+    1. Key ID
+    2. Value ID or None -  Deprecated, but left for backward compatibility with old data. Should be None all the time.
+    3. Value Template or None – By business logic it should be always present, but it might be None if a legacy data.
+    It means only [key_id, None, template] should be present.
+    // TODO: refactor to use just regular DB fields for dynamic label config.
+    """
+    DynamicLabelsEntryDB = tuple[str, str | None, str | None]
+    DynamicLabelsConfigDB = list[DynamicLabelsEntryDB] | None
+    alert_group_labels_custom: DynamicLabelsConfigDB = models.JSONField(null=True, default=None)
```
[the remark this answers is not in the document]
Comment (Konstantinov-Innokentii) [reply engine/apps/alerts/models/alert_receive_channel.py:315]: There is some further refactoring planned on that piece

File: engine/apps/api/serializers/alert_receive_channel.py [hunk 6]
```diff
@@ -206,7 +208,7 @@ def to_representation(cls, instance: AlertReceiveChannel) -> IntegrationAlertGro
     @staticmethod
     def _custom_labels_to_internal_value(
         custom_labels: AlertGroupCustomLabelsAPI,
```
[the remark this answers is not in the document]
Comment (Konstantinov-Innokentii) [reply engine/apps/api/serializers/alert_receive_channel.py:210]: Yep, we should, it will be in separate PRs to keep bugfix tight.

**Pull request merged**